File: QuantConnect/Strategy_3.py


# Trend following strategy using MACD with signal line cross entry and opposite signal or trailing stop exit
from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import *
from QuantConnect.Algorithm import QCAlgorithm
from QuantConnect.Orders import *
from QuantConnect.Securities import *


class TrendFollowingMACD(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2010, 1, 1)
        self.SetEndDate(2021, 1, 1)
        self.SetCash(100000)
        
        self.symbol = "SPY"
        self.atrPeriod = 14
        self.atrMultiplier = 1
        
        # MACD indicator setup
        self.fastPeriod = 12
        self.slowPeriod = 26
        self.signalPeriod = 9
        
        # Set up indicator for the SPY symbol
        self.spy = self.AddEquity(self.symbol, Resolution.Daily).Symbol
        
        # Add indicators to algorithm using helper methods available in QCAlgorithm base class
        macd = self.MACD(self.spy, self.fastPeriod, self.slowPeriod, self.signalPeriod, MovingAverageType.Exponential)
        
        # Use the MACD signal line as our entry indicator for long positions only (filtering out short signals)
        macdSignalLineCrossOver = CrossAbove(macd.Signal, macd.Fast, 1)
        
        # Use the opposite signal or trailing stop as our exit criteria
        macdSignalLineCrossUnder = CrossUnder(macd.Signal, macd.Fast, 1)
        
        # Set up trailing stop using ATR (Average True Range) to calculate stop distance
        atr = AverageTrueRange(self.atrPeriod)
        
    def OnData(self, data):
            if not (data.ContainsKey(self.spy) and data[self.spy] and data[self.spy].Price):
                return
            
            if not (self.IsWarmingUp):
                price = data[self.spy].Price
                
                if not (self.Portfolio.Invested) and macdSignalLineCrossOver:
                    # Enter long position at market price and set stop loss based on ATR * multiplier from current price
                    stopPrice = price - (self.atrMultiplier * atr.Current.Value)
                    quantity = int(self.Portfolio.Cash / price)
                    orderTicket = self.MarketOrder(self.spy, quantity)
                    orderTicket.StopMarket(stopPrice)

                elif macdSignalLineCrossUnder or (self.Portfolio.Invested and price < (stopPrice + atr.Current.Value)):
                    # Exit long position at market price and cancel any open stop loss orders 
                    orderTicket.StopMarket(stopPrice).Cancel()
                    quantity = int(self.Portfolio[self.spy].Quantity * -1)
                    orderTicket = self.MarketOrder(self.spy, quantity)

File: QuantConnect/Strategy_3.cs


# TrendFollowing_MACD.cs

using System;
using System.Linq;
using QuantConnect.Data.Market;
using QuantConnect.Indicators;

namespace QuantConnect.Algorithm
{
    public class TrendFollowing_MACD : QCAlgorithm
    {
        private MovingAverageConvergenceDivergence _macd;
        private ExponentialMovingAverage _signal;
        private RelativeStrengthIndex _rsi;

        private string _symbol = "SPY";
        private decimal _atrStop = 1m;

        public override void Initialize()
        {
            SetStartDate(2010, 1, 1);
            SetEndDate(DateTime.Now.Date.AddDays(-1));
            SetCash(100000);
            AddEquity(_symbol, Resolution.Daily);

            _macd = MACD(_symbol, 12, 26, 9, MovingAverageType.Exponential);
            _signal = EMA(_macd.Signal, 9);
            _rsi = RSI(_symbol, 14);

            // set warmup period to ensure indicators are ready
            int warmupPeriod = Math.Max(_macd.Signal.WarmUpPeriod + _signal.WarmUpPeriod, _rsi.WarmUpPeriod);
            SetWarmUp(TimeSpan.FromDays(warmupPeriod));
        }

        public override void OnData(Slice data)
        {
            if (!Portfolio.Invested)
            {
                // enter long position on signal line cross above MACD line
                if (_macd.Current.Value > _signal.Current.Value && 
                    _macd.Current.Value < _macd.Previous.Value &&
                    !_rsi.IsOverbought())
                {
                    decimal stopPrice = data[_symbol].High * (1m - (_atrStop / 100m));
                    SetHoldings(_symbol, 1m);
                    SetStopLoss(_symbol, stopPrice);
                }
            }
            else
            {
                // exit long position on signal line cross below MACD line or trailing stop hit
                if (_macd.Current.Value < _signal.Current.Value ||
                    Portfolio[_symbol].UnrealizedProfitPercent < -_atrStop)
                {
                    Liquidate();
                }
                else
                {
                    // update stop loss based on current ATR value
                    decimal atrValue = ATR(_symbol, 14).Current.Value;
                    decimal newStopPrice = data[_symbol].High * (1m - ((atrValue * _atrStop) / 100m));
                    UpdateStopLoss(_symbol, newStopPrice);
                }
            }
        }
    }

    public static class Extensions
    {
        public static bool IsOverbought(this RelativeStrengthIndex rsi)
        {
            return rsi > 70m;
        }
    }
}